### PR TITLE
Add new Dataset for training joint ML model

### LINKED ISSUE #59: Check our models are getting this configuration of data inputs
## General
- [ ] The target should be future GSP power and, optionally, future satellite imagery 

## The model should receive these as inputs
- [ ] An embedding of the GSP ID (also see https://github.com/openclimatefix/nowcasting_dataset/issues/451)
- [ ] Historical data from all individual PV systems within the region of interest (which almost certainly will be available at inference time and is probably important so the model can see how much sunlight each cloud is letting through!), probably including:
    - [ ] An embedding of each PV system's ID (so the model can learn which PV systems to trust)?
    - [ ] The spatial and temporal location of each PV reading

## The model should NOT receive these as inputs :slightly_smiling_face:
- [ ] Triple-check that we're not accidentally giving the model future GSP power as an input :slightly_smiling_face:  (mistakes like this are all too easy to make!)
- [ ] Historical GSP power data (because that won't be available at inference time!)
- [ ] Future satellite images :slightly_smiling_face: 
- [ ] Future data from individual PV systems

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request

## Description

Adds another dataset that outputs satellite and GSP data, and ensures that the model is getting the correct inputs and outputs.

This PR:

- [x] Adds a new Dataset that splits the inputs into past/future timesteps
- [x] Splits the position encodings into future and past parts, appending the past ones to the inputs
- [x] Removes past GSP data
- [x] Returns a target of future GSP data and, optionally, future satellite imagery
- [x] Returns inputs of past data from various input modalities, as well as future timesteps of NWP data

Fixes #59 

## How Has This Been Tested?

Unit tests

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: nowcasting_dataloader/datasets.py [hunk 1]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+class SatFlowDataset(NetCDFDataset):
+    """SatFlow dataset for filtering and splitting up output from NetCDFDataset properly"""
+
+    def __init__(
+        self,
+        n_batches: int,
+        src_path: str,
+        tmp_path: str,
+        configuration: Configuration,
+        cloud: str = "gcp",
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:199]: In the future, it'd be good to remove the concept of `cloud` from `nowcasting_dataloader`: #61

(but definitely don't worry about that for this PR!)

File: nowcasting_dataloader/datasets.py [hunk 2]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+        required_keys: Union[Tuple[str], List[str]] = None,
+        history_minutes: Optional[int] = None,
+        forecast_minutes: Optional[int] = None,
+        normalize: bool = False,
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:203]: should `normalize` default to `True`?

File: nowcasting_dataloader/datasets.py [hunk 3]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+            forecast_minutes: How many future minutes of data to use, if reducing the amount
+                of forecast time
+            configuration: configuration object
+            cloud: which cloud is used, can be "gcp", "aws" or "local".
+            normalize: normalize the batch data
+            add_position_encoding: Whether to add position encoding or not
+            add_satellite_target: Whether to add future satellite imagery to the target or not
+            add_hrv_satellite_target: Whether to add future HRV satellite imagery to the target
+        """
+        super().__init__(
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:230]: with this many arguments, I'd definitely recommend using named params e.g. `(n_batches=n_batches, ...)`....

File: nowcasting_dataloader/datasets.py [hunk 4]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+
+    def __getitem__(self, batch_idx: int) -> Tuple[dict, dict]:
+        """
+        Satflow extension for the dataloader
+
+        Args:
+            batch_idx: Batch ID to load
+
+        Returns:
+            Tuple of dicts of torch.Tensors holding the data
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:258]: Please document what the two dicts in the tuple hold the input and the target, respectively.

File: nowcasting_dataloader/datasets.py [hunk 5]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+            past_hrv_satellite_data = batch["hrvsatellite"]["data"][
+                :, :, : self.current_timestep_index
+            ]
+            x["hrvsatellite"] = past_hrv_satellite_data
+        if len(batch["pv"].get(PV_YIELD, [])) > 0:
+            past_pv_data = batch["pv"][PV_YIELD][:, :, : self.current_timestep_index]
+            x[PV_YIELD] = past_pv_data
+            x[PV_SYSTEM_ID] = batch["pv"][PV_SYSTEM_ID]
+        if len(batch["nwp"].get("data", [])) > 0:
+            # We can give future NWP too, as that will be available
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:277]: good comment!

File: nowcasting_dataloader/datasets.py [hunk 6]
```diff
@@ -179,6 +187,182 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+                if self.add_hrv_satellite_target:
+                    x["hrv_" + SATELLITE_DATA + "_query"] = x.pop("hrvsatellite_query")
+            if len(x.get(TOPOGRAPHIC_DATA, [])) > 0:
+                x = self.add_encodings(x, TOPOGRAPHIC_DATA, batch, 0, False)
+            if len(x.get(NWP_DATA, [])) > 0:
+                x[NWP_DATA] = torch.cat(
+                    [x[NWP_DATA], batch[NWP_DATA + "_position_encoding"]], dim=1
+                )
+            if len(x.get(PV_YIELD, [])) > 0:
+                x = self.add_encodings(x, PV_YIELD, batch, self.current_timestep_index, False)
```
Comment (JackKelly) [inline nowcasting_dataloader/datasets.py:323]: Can we also add an encoding of the geographical location of each PV system?  (Maybe as a separate PR?) 

(not super-super urgent as the model might be able to infer to location of each PV system)

(Sorry if this code already does this and I've missed it!)

**Review by JackKelly — APPROVED** [review]
Awesome work!  So exciting to see this all coming together!

File: nowcasting_dataloader/datasets.py (hunk 6 above)
Comment (jacobbieker) [reply nowcasting_dataloader/datasets.py:323]: This already does it! It takes the coordinates of the GSP and PV systems and adds them in NetCDFDataset position encodings.

File: nowcasting_dataloader/datasets.py [hunk 7]
```diff
@@ -179,6 +187,183 @@ def __getitem__(self, batch_idx: int) -> dict:
...
+
+        Returns:
+            Tuple of dicts of torch.Tensors holding the data, with the first one being the inputs
+            and the second being the target
+        """
+        batch = super().__getitem__(batch_idx)
+        x = {}
+        target = {}
+        # Need to partition out past and future sat images here, along with the rest of the data
+        if len(batch["satellite"].get("data", [])) > 0:
```
Comment (peterdudfield) [inline nowcasting_dataloader/datasets.py:265]: I wonder if you want to move it back to a `Batch` class, or any option to return it like that. 

That'll make things like 'batch["nwp"]["data"]' go to batch.nwp.data whcih is think is better. 

Could put this perhaps as a TODO

File: tests/test_satflow_dataset.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            "hrv_sat_data",
+        ]:
+            assert k in x.keys()
+            assert type(x[k]) == torch.Tensor
+
+        for k in ["gsp_yield", "gsp_id", "sat_data", "hrv_sat_data"]:
+            assert k in y.keys()
+            assert type(y[k]) == torch.Tensor
+        # Make sure file isn't deleted!
+        assert os.path.exists(os.path.join(DATA_PATH, "metadata/000000.nc"))
```
Comment (peterdudfield) [inline tests/test_satflow_dataset.py:147]: perhaps move this to nwp, as I'm sure well get rid of metadata soon

**Review by peterdudfield — APPROVED** [review]

File: nowcasting_dataloader/datasets.py (hunk 7 above)
Comment (jacobbieker) [reply nowcasting_dataloader/datasets.py:265]: Yeah, this would involve putting the position encodings into the `Batch` or `BatchML` as well, so probably put it as a TODO for now?
